lib/IDE/Refactoring.cpp

```diff
@@ -1667,6 +1667,366 @@ bool RefactoringActionConvertStringsConcatenationToInterpolation::performChange(
   return false;
 }
 
+struct ExpandableTernaryExprInfo {
+
+  ExpandableTernaryExprInfo() {}
+
+  swift::IfExpr *getIf() {
+    if (Binding && Binding->getNumPatternEntries() == 1)
+      return dyn_cast<IfExpr>(Binding->getInit(0));
+    if (Assign)
+      return dyn_cast<IfExpr>(Assign->getSrc());
+    return nullptr;
+  }
+
+  swift::SourceRange getNameRange() {
+    if (Binding) {
+      if (auto pattern = getNamePattern()) {
+        return pattern->getSourceRange();
+      }
+    }
+    if (Assign) {
+      if (auto dest = Assign->getDest()) {
+        return dest->getSourceRange();
+      }
+    }
+    return swift::SourceRange();
+  }
+
+  swift::Type getType() {
+    if (Binding) {
+      if (auto pattern = getNamePattern()) {
+        return pattern->getType();
+      }
+    }
+    return swift::Type();
+  }
+
+  bool shouldDeclareNameAndType() {
+    return !getType().isNull();
+  }
+
+  bool isValid() {
+    bool isBindingDecl =  Binding && !Assign;
+    bool isAssignExpr  = !Binding &&  Assign;
+
+    //Ensure info is PatternBindingDecl or AssignExpr
+    if (!isBindingDecl && !isAssignExpr)
+      return false;
+    if (Binding && Binding->getNumPatternEntries() != 1)
+      return false;
+
+    //Ensure all public properties are non-nil and valid
+    if (!getIf() || !getNameRange().isValid())
+      return false;
+    if (shouldDeclareNameAndType() && getType().isNull())
+      return false;
+
+    //All validations passed
+    return true;
+  }
+
+  void setBinding(swift::PatternBindingDecl *NewBinding) {
+    Binding = NewBinding;
+  }
+
+  void setAssign(swift::AssignExpr *NewAssign) {
+    Assign = NewAssign;
+  }
+
+private:
+
+  swift::Pattern *getNamePattern() {
+    if (Binding && Binding->getNumPatternEntries() == 1) {
+      if (auto pattern = Binding->getPattern(0)) {
+        if (auto typedPattern = dyn_cast<TypedPattern>(pattern)) {
+          pattern = typedPattern->getSubPattern();
+        }
+        return pattern;
+      }
+    }
+    return nullptr;
+  }
+
+  swift::AssignExpr *Assign = nullptr;
+  swift::PatternBindingDecl *Binding = nullptr;
+};
+
+ExpandableTernaryExprInfo
+findExpandableTernaryExpression(ResolvedRangeInfo Info) {
+
+  auto notFound = ExpandableTernaryExprInfo();
+
+  if (Info.Kind != RangeKind::SingleDecl
+      && Info.Kind != RangeKind:: SingleExpression)
+    return notFound;
+
+  if (Info.ContainedNodes.size() != 1)
+    return notFound;
+
+  ExpandableTernaryExprInfo Target;
+
+  if (auto D = Info.ContainedNodes[0].dyn_cast<Decl*>())
+    Target.setBinding(dyn_cast<PatternBindingDecl>(D));
```

Comment:
Once we separate the logic for PatternBindingDecl and AssignExpr, we can use constructors to pass the decl or the expr without using `setX()` functions.